server/vulnerabilities/cpe.go

```diff
@@ -140,67 +115,391 @@ func cleanAppName(appName string) string {
-	targetSW := ""
-	switch software.Source {
-	case "apps":
-		targetSW = "macos"
-	case "python_packages":
-		targetSW = "python"
-	case "chrome_extensions":
-		targetSW = "chrome"
-	case "firefox_addons":
-		targetSW = "firefox"
-	case "safari_extensions":
-		targetSW = "safari"
-	case "deb_packages":
-	case "portage_packages":
-	case "rpm_packages":
-	case "npm_packages":
-		targetSW = `"node.js"`
-	case "atom_packages":
-	case "programs":
-		targetSW = `"windows*"`
-	case "ie_extensions":
-	case "chocolatey_packages":
+var nonAlphaNumeric = regexp.MustCompile(`[^a-zA-Z0-9]+`)
+
+// sanitizeMatch sanitizes the search string for sqlite fts queries. Replaces all non alpha numeric characters with spaces.
+func sanitizeMatch(s string) string {
+	return nonAlphaNumeric.ReplaceAllString(s, " ")
+}
+
+var sanitizeVersionRe = regexp.MustCompile(`[^a-zA-Z0-9_-]+`)
+
+// sanitizeVersion attempts to sanitize versions and attempt to make it dot separated.
+// Eg Zoom reports version as "5.11.1 (8356)". In the NVD CPE dictionary it should be 5.11.1.8356.
+func sanitizeVersion(version string) string {
+	parts := sanitizeVersionRe.Split(version, -1)
+	return strings.Trim(strings.Join(parts, "."), ".")
+}
+
+const cpeTranslationsFilename = "cpe_translations.json"
+
+func loadCPETranslations(path string) (CPETranslations, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	var translations CPETranslations
+	if err := json.NewDecoder(f).Decode(&translations); err != nil {
+		return nil, fmt.Errorf("decode json: %w", err)
+	}
+
+	return translations, nil
+}
+
+// DownloadCPETranslations downloads the CPE translations to the given vulnPath. If cpeTranslationsURL is empty, attempts to download it
+// from the latest release of github.com/fleetdm/nvd. Skips downloading if CPE translations is newer than the release.
+func DownloadCPETranslations(vulnPath string, client *http.Client, cpeTranslationsURL string) error {
+	path := filepath.Join(vulnPath, cpeTranslationsFilename)
+
+	if cpeTranslationsURL == "" {
+		release, err := GetLatestNVDRelease(client)
+		if err != nil {
+			return err
+		}
+		stat, err := os.Stat(path)
+		switch {
+		case errors.Is(err, os.ErrNotExist):
+			// okay
+		case err != nil:
+			return err
+		default:
+			if stat.ModTime().After(release.CreatedAt.Time) {
+				// file is newer than release, do nothing
+				return nil
+			}
+		}
+
+		for _, asset := range release.Assets {
+			if cpeTranslationsFilename == asset.GetName() {
+				cpeTranslationsURL = asset.GetBrowserDownloadURL()
+				break
+			}
+		}
+		if cpeTranslationsURL == "" {
+			return errors.New("failed to find cpe translations in nvd release")
+
+		}
+	}
+
+	u, err := url.Parse(cpeTranslationsURL)
+	if err != nil {
+		return err
+	}
+	if err := download.Download(client, u, path); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// RegexpCache caches compiled regular expressions. Not safe for concurrent use.
```

Comment:
No need to export in this iteration: 
- only used internally
- such functionality doesn't really belong to `vulnerabilities`

Reply:
fixed